airflow/utils/log/secrets_masker.py

```diff
@@ -0,0 +1,221 @@
+            for k, v in record.__dict__.items():
+                if k in self._record_attrs_to_ignore:
+                    continue
+                record.__dict__[k] = self.redact(v)
+            if record.exc_info:
+                exc = record.exc_info[1]
+                # I'm not sure if this is a good idea!
+                exc.args = (self.redact(v) for v in exc.args)
+        record.__dict__[self.ALREADY_FILTERED_FLAG] = True
+
+        return True
+
+    def _redact_all(self, item: "RedactableItem") -> "RedactableItem":
+        if isinstance(item, dict):
+            return {dict_key: self._redact_all(subval) for dict_key, subval in item.items()}
+        elif isinstance(item, str):
+            return '***'
+        elif isinstance(item, (tuple, set)):
+            # Turn set in to tuple!
+            return tuple(self._redact_all(subval) for subval in item)
+        elif isinstance(item, (Iterable, list)):
+            return list(self._redact_all(subval) for subval in item)
+        else:
+            return item
+
+    # pylint: disable=too-many-return-statements
+    def redact(self, item: "RedactableItem", name: str = None) -> "RedactableItem":
+        """
+        Redact an any secrets found in ``item``, if it is a string.
+
+        If ``name`` is given, and it's a "sensitve" name (see
+        :func:`should_hide_value_for_key`) then all string values in the item
+        is redacted.
+
+        """
+        if name and should_hide_value_for_key(name):
+            return self._redact_all(item)
+
+        if isinstance(item, dict):
+            return {dict_key: self.redact(subval, dict_key) for dict_key, subval in item.items()}
+        elif isinstance(item, str):
+            if self.replacer:
+                # We can't replace specific values, but the key-based redacting
+                # can still happen, so we can't short-circuit, we need to walk
+                # the strucutre.
+                return self.replacer.sub('***', item)
+            return item
+        elif isinstance(item, (tuple, set)):
+            # Turn set in to tuple!
+            return tuple(self.redact(subval) for subval in item)
+        elif isinstance(item, Iterable):
+            return list(self.redact(subval) for subval in item)
+        else:
+            return item
+
+    # pylint: enable=too-many-return-statements
+
+    def add_mask(self, secret: Union[str, dict, Iterable], name: str = None):
+        """Add a new secret to be masked to this filter instance."""
+        if isinstance(secret, dict):
+            for k, v in secret.items():
+                self.add_mask(v, k)
+        elif isinstance(secret, str):
+            pattern = re.escape(secret)
+            if pattern not in self.patterns and (not name or should_hide_value_for_key(name)):
+                self.patterns.add(pattern)
+                self.replacer = re.compile('|'.join(self.patterns))
+        elif isinstance(secret, collections.abc.Iterable):
```

Comment:
I would suggest that we use try-except for checking if secret is an iterator. collections.abc.Iterable doesn't detect iterators unless they implement __iter__ while this does:
```python
try:
    iter(myobject)
except TypeError:
    print('not iterator')
```
Found this from https://stackoverflow.com/a/36407550/3477338

Reply:
That was sort of by design, since we turn the output in to a list always, so we don't want to redact everything.

I'm not sure that supporting Iterable is good

Reply:
@ashb - your lackluster feelings about  Iterable support was right I think https://github.com/apache/airflow/issues/16148#issuecomment-855240855 :)